### ISSUE TITLE
Reports false type errors on macro, but actually compiles without errors

### ISSUE THREAD

**Issue by lzyor** [rust-lang/rust-analyzer#21837]
**rust-analyzer version**: 0.3.2828-standalone

**rustc version**: rustc 1.90.0 (1159e78c4 2025-09-14) ~ rustc 1.94.0 (4a4ef493e 2026-03-02)

**editor or extension**: VSCode 1.109.5

**code snippet to reproduce**:
```rust
use ambassador::{Delegate, delegatable_trait};

#[delegatable_trait]
pub trait Shout {
    fn shout(&self, input: &str) -> String;
}

pub struct Cat;

impl Shout for Cat {
    fn shout(&self, input: &str) -> String {
        format!("{} - meow!", input)
    }
}

pub struct Dog;

impl Shout for Dog {
    fn shout(&self, input: &str) -> String {
        format!("{} - wuff!", input)
    }
}

#[derive(Delegate)]
#[delegate(Shout)]
pub enum Animal {
    Cat(Cat),
    Dog(Dog),
}

pub fn main() {
    let foo_animal = Animal::Cat(Cat);
    println!("{}", foo_animal.shout("BAR"));
}
```

Simply run the example code from the `ambassador` crate documentation: https://crates.io/crates/ambassador

rust-analyzer reports errors:

```text
---> #[derive(Delegate)]
expected &{unknown}, found Dog rust-analyzer(E0308)
expected &{unknown}, found Cat rust-analyzer(E0308)
```

However, there are actually no errors, and the macro-expanded code compiles successfully.

**Comment by ChayimFriedman2** [comment]
Likely related to order of name resolution.

**Comment by AayushMainali-Github** [comment]
@ChayimFriedman2  I opened a PR for this with a fix and regression test. Would appreciate a look when
  you get a chance.

**Comment by zirconium-n** [comment]
I hit this problem as well. Here's some more info.

`#[delegatable_trait]` expands to:

```rust
#[macro_export]
#[doc(hidden)]
macro_rules! _ambassador_impl_Shout {
    (body_struct(< > , $ty:ty, $field_ident:tt)) => {
        ambassador_impl_Shout!{
            body_struct(< > , $ty,($field_ident),($field_ident),($field_ident))
        }
    };
    (body_struct(< > , $ty:ty,($($ident_owned:tt)*),($($ident_ref:tt)*),($($ident_ref_mut:tt)*))) => {
        #[inline]#[allow(unused_braces)]fn shout(&self,input: &str)->String {
            Shout::shout:: < >(ambassador_impl_Shout!(resolve_path("target_ref was not specified but was needed", & ,self. $($ident_ref)*)),input)
        }
    };
    (resolve_path($err:literal, $(& $(mut )?)? , $s:ident. $field:ident$($t:tt)+)) => {
        $s. $field$($t)*
    };
    (resolve_path($err:literal, $(& $(mut $($emp:lifetime)?)?)? , $s:ident. $field:tt)) => {
        $(& $(mut $($emp)?)?)? $s. $field
    };
    (resolve_path($err:literal, $(& $(mut )?)? , $s:ident.)) => {
        compile_error!{
            $err
        }
    };
    (body_enum(< > , $ty:ty,($($other_tys:ty), *),($($variants:path), +))) => {
        #[inline]#[allow(unused_braces)]fn shout(&self,input: &str)->String {
            match self {
                $($variants(inner) => Shout::shout:: < >(inner,input)), *
            }
        }
    };
    (body_self(< >)) => {
        #[inline]#[allow(unused_braces)]#[deny(unconditional_recursion)]fn shout(&self,input: &str)->String {
            self.shout:: < >(input)
        }
    };
    (use_assoc_ty_bounds) => {
        #[doc(hidden)]#[allow(non_camel_case_types)]pub trait MatchShout<ambassador_X:Shout< >> :Shout< >{}
        #[allow(non_camel_case_types)]impl <ambassador_X:Shout< > ,ambassador_Y:Shout< > >MatchShout<ambassador_X>for ambassador_Y{}

    };
}
#[doc(inline)]
#[doc = concat!("A macro to be used by [`ambassador::Delegate`] to delegate [`",stringify!(Shout),"`]")]
pub use _ambassador_impl_Shout as ambassador_impl_Shout;
#[doc(hidden)]
#[allow(non_snake_case)]
pub mod ambassador_impl_Shout {}
```
And if I inline this and remove `#[delegatable_trait]`. The error goes away.

-------

```rust
#[derive(Delegate)]
#[delegate(Shout)]
```
expands to following:
```rust
#[allow(non_snake_case)]
mod ambassador_module_Shout_for_Animal {
    use super::*;
    ambassador_impl_Shout! {
        use_assoc_ty_bounds
    }
    impl Shout for Animal
    where
        Dog: Shout,
        Cat: MatchShout<Dog>,
    {
        ambassador_impl_Shout! {
            body_enum(< > ,Dog,(Cat),(Animal::Cat,Animal::Dog))
        }
    }
}
```
The error remains if I inline this, but if I continue to inline 
```rust
ambassador_impl_Shout! {
    body_enum(< > ,Dog,(Cat),(Animal::Cat,Animal::Dog))
}
```
into
```rust
#[inline]
#[allow(unused_braces)]
fn shout(&self,input: &str) -> String {
    match self {
        Animal::Cat(inner) => Shout::shout:: < >(inner,input),
        Animal::Dog(inner) => Shout::shout:: < >(inner,input)
    }
}
```
The error goes away as well.

Apparently the error came from the match statement, but i'm not sure. If I replace `match self` with `match *self` I get the same error, but that's not what's being expanded into anyway.

**Comment by zirconium-n** [comment]
Seems related to interaction between macro hygiene and `self` indentifier.
